text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1471 @@
+  // discouraged case is having a string label and a numeric label of the same
+  // name on a sample.  Again, possible to express, but should not be used.
+  // [deprecated, superseded by attributes]
+  repeated Label label = 3;
+
+  // locations_start_index along with locations_length refers to to a slice of locations in Profile.location. Supersedes location_index.
+  uint64 locations_start_index = 4;
+
+  // locations_length along with locations_start_index refers to a slice of locations in Profile.location. Supersedes location_index.
+  uint64 locations_length = 5;
+
+  // Timestamps associated with Sample represented in ms. These timestamps are expected to fall within the Profile's time range. [optional]
+  repeated uint64 timestamps = 6;
+
+  // References to attributes in Profile.attribute_table. [optional]
+  repeated uint64 attributes = 7;
+
+  // A 128bit id that uniquely identifies this stacktrace, globally. Index into string table. [optional]
+  uint32 stacktrace_id_index = 2;
+}
+
+message Label {
+  int64 key = 1;   // Index into string table
+
+  // At most one of the following must be present
+  int64 str = 2;   // Index into string table
+  int64 num = 3;
+
+  // Should only be present when num is present.
+  // Specifies the units of num.
+  // Use arbitrary string (for example, "requests") as a custom count unit.
+  // If no unit is specified, consumer may apply heuristic to deduce the unit.
+  // Consumers may also  interpret units like "bytes" and "kilobytes" as memory
+  // units and units like "seconds" and "nanoseconds" as time units,
+  // and apply appropriate unit conversions to these.
+  int64 num_unit = 4;  // Index into string table
+}
+
+message Mapping {
+  // Unique nonzero id for the mapping. [deprecated]
+  uint64 id = 1;
+  // Address at which the binary (or DLL) is loaded into memory.
+  uint64 memory_start = 2;
+  // The limit of the address range occupied by this mapping.
+  uint64 memory_limit = 3;
+  // Offset in the binary that corresponds to the first mapped address.
+  uint64 file_offset = 4;
+  // The object this entry is loaded from.  This can be a filename on
+  // disk for the main binary and shared libraries, or virtual
+  // abstractions like "[vdso]".
+  int64 filename = 5;  // Index into string table
+  // A string that uniquely identifies a particular program version
+  // with high probability. E.g., for binaries generated by GNU tools,
+  // it could be the contents of the .note.gnu.build-id field.
+  int64 build_id = 6;  // Index into string table
+
+  // The following fields indicate the resolution of symbolic info.
+  bool has_functions = 7;
+  bool has_filenames = 8;
+  bool has_line_numbers = 9;
+  bool has_inline_frames = 10;
+
+  // A string that uniquely identifies a particular binary with high probability.
+  // It is meant to work around deficiencies in the GNU Build ID that may compromise
+  // uniqueness / correlatability. Index into string table.
+  uint32 file_id_index = 11;
```

Comment:
Could we simply use the build_id field for this? The two fields are very close in semantics so I'm not sure it's great to carry both of them forward forever...

Reply:
If the matter is being able to distinguish the semantics (GNU build ID vs alternative hash), maybe we could instead add an enum that tells the exact meaning of the build_id field with zero value enum member for the current semantics? I understand that build_id maybe non-ideal name, but perhaps it's good enough?

Reply:
I know we've discussed this on Dec 14, but not sure what we've decided on here. @aalexand and @thomasdullien,  what do you guys think?

Reply:
I think what we've decided and what I'd prefer is that we keep the pprof's build_id field as is but also add a build_id_kind enum field with BuildIdKind enum that has members like `BUILD_ID_LINKER = 0` corresponding to the current semantics and `BUILD_ID_BINARY_HASH = 1` for the proposed new schema. In the future it can be extended to other hashing approaches if and as necessary. The only downside to this I see is imperfect name (one could argue that file_id is cleaner) but I think backward compatibility trumps the perfection in this case.